### ISSUE TITLE
Delete docs could cause missing undeleted docs in ForestDB view index 

### ISSUE THREAD

**Issue by pasin** [couchbase/couchbase-lite-ios#1082]
I have found this issue (https://github.com/couchbaselabs/ToDoLite-iOS/issues/57) when I test ToDoLite-iOS with ForestDB Storage . I was able to reproduce this with a unit test as below. The issue only happens if deleting doc2 but not doc1 or doc3.

```
- (void) test23_ViewWithDocDeletion {
    CBLView* view = [db viewNamed: @"vu"];
    Assert(view);
    [view setMapBlock: MAPBLOCK({
        if ([doc[@"type"] isEqualToString: @"task"]) {
            id date = doc[@"created_at"];
            NSString* listID = doc[@"list_id"];
            emit(@[listID, date], doc);
        }
    }) version: @"1"];
    Assert(view.mapBlock != nil);
    AssertEq(view.totalRows, 0u);
    
    NSString* listId = @"list1";
    
    // Create 3 documents:
    CBLDocument* doc1 = [self createDocumentWithProperties:
                         @{@"_id": @"doc1",
                           @"type": @"task",
                           @"created_at": [CBLJSON JSONObjectWithDate: [NSDate date]],
                           @"list_id": listId}];
    Assert(doc1);
    CBLDocument* doc2 = [self createDocumentWithProperties:
                         @{@"_id": @"doc2",
                           @"type": @"task",
                           @"created_at": [CBLJSON JSONObjectWithDate: [NSDate date]],
                           @"list_id": listId}];
    Assert(doc2);
    CBLDocument* doc3 = [self createDocumentWithProperties:
                         @{@"_id": @"doc3",
                           @"type": @"task",
                           @"created_at": [CBLJSON JSONObjectWithDate: [NSDate date]],
                           @"list_id": listId}];
    Assert(doc3);
    
    // Check query result:
    CBLQuery* query = [view createQuery];
    query.descending = YES;
    query.startKey = @[listId, @{}];
    query.endKey = @[listId];
    
    CBLQueryEnumerator* rows = [query run: NULL];
    AssertEq(rows.count, 3u);
    AssertEqual([rows rowAtIndex:0].documentID, doc3.documentID);
    AssertEqual([rows rowAtIndex:1].documentID, doc2.documentID);
    AssertEqual([rows rowAtIndex:2].documentID, doc1.documentID);
    
    // Delete doc2:
    Assert(doc2);
    NSError* error;
    Assert([doc2 deleteDocument: &error]);
    
    // Check query result:
    rows = [query run: NULL];
    AssertEq(rows.count, 2u);
    AssertEqual([rows rowAtIndex:0].documentID, doc3.documentID);
    AssertEqual([rows rowAtIndex:1].documentID, doc1.documentID);
}
```

**Comment by snej** [comment]
If the queries run in ascending order instead, the test passes. So it's triggered by the descending iteration.

**Comment by snej** [comment]
The listing above contains `Assert([doc3 deleteDocument: &error]);` but I believe this is a mistake; it's supposed to delete `doc2`, and that's what was given in the original listing Pasin showed me.

**Comment by pasin** [comment]
That's copy & paste mistake, Just change it to doc2.

**Comment by snej** [comment]
I'm stepping through the CBForest index enumerator code, and I think the bug is down in ForestDB itself: after the `fdb_iterator` returns the first row (doc3), it stops without returning the row for doc1.

**Comment by snej** [comment]
I have a workaround that causes the unit test to pass. Need to run the full CBForest and CBL unit test suite.

The workaround is in the DocEnumerator class: In a descending iteration, it now avoids setting ForestDB's FDB_ITR_NO_DELETES option. Instead it checks each document's metadata, and if its deleted flag is set, it skips it.

**Comment by snej** [comment]
Filed a ForestDB issue, [MB-17785](https://issues.couchbase.com/browse/MB-17785).

**Comment by snej** [comment]
Fixed on release/1.2 branch.
